File: typescript/modules/libs/workspace/derivations/apps/src/lib/rules/postal-address-display-name.rule.ts
import { stripIndents, oneLine, inlineLists } from 'common-tags';

import { Dependency, RoleType } from '@allors/workspace/meta/system';
import { IRule } from '@allors/workspace/domain/system';
import { M } from '@allors/workspace/meta/default';
import { PostalAddress } from '@allors/workspace/domain/default';

export class PostalAddressDisplayNameRule implements IRule<PostalAddress> {
  roleType: RoleType;
  dependencies: Dependency[];

  constructor(m: M) {
    const { dependency: d } = m;

    this.roleType = m.PostalAddress.DisplayName;

    this.dependencies = [d(m.PostalAddress, (v) => v.Country)];
  }

  derive(postalAddress: PostalAddress) {
    postalAddress.DisplayName = stripIndents`
      ${[postalAddress.Address1, postalAddress.Address2, postalAddress.Address3].filter((v) => v).map((v) => oneLine`${v}`)}
      ${inlineLists`${[postalAddress.PostalCode, postalAddress.Locality].filter((v) => v)}`}
      ${postalAddress.Country?.Name ?? ''}
      `;
  }
}

File: dotnet/Apps/Database/Domain/apps/rules/relations/postaladdressdisplaynamerule.cs
// <copyright file="Domain.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Database.Domain
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Database.Derivations;
    using Meta;
    using Derivations.Rules;

    public class PostalAddressDisplayNameRule : Rule
    {
        public PostalAddressDisplayNameRule(MetaPopulation m) : base(m, new Guid("5922677a-6e7d-4ada-bd7c-0104edb278d8")) =>
            this.Patterns = new Pattern[]
            {
                m.PostalAddress.RolePattern(v => v.Address1),
                m.PostalAddress.RolePattern(v => v.Address2),
                m.PostalAddress.RolePattern(v => v.Address3),
                m.PostalAddress.RolePattern(v => v.PostalCode),
                m.PostalAddress.RolePattern(v => v.Locality),
                m.PostalAddress.RolePattern(v => v.Country),
            };

        public override void Derive(ICycle cycle, IEnumerable<IObject> matches)
        {
            foreach (var @this in matches.Cast<PostalAddress>())
            {
                var array = new string[] {@this.Address1, @this.Address2, @this.Address3, @this.PostalCode, @this.Locality, @this.Country?.Name };
                @this.DisplayName = string.Join(", ", array.Where(s => !string.IsNullOrEmpty(s)));
            }
        }
    }
}
